secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/dcr/request/RegistrationRequestEntityValidatorFilter.java

```diff
@@ -0,0 +1,150 @@
+    private final TrustedDirectoryService trustedDirectoryService;
+    private final RegistrationRequestBuilder registrationRequestBuilder;
+    private final JwtDecoder jwtDecoder;
+    private final AcceptHeaderSupplier acceptHeaderSupplier;
+    private final ResponseFactory responseFactory;
+
+    public RegistrationRequestEntityValidatorFilter(RegistrationRequestEntitySupplier registrationEntitySupplier,
+            AcceptHeaderSupplier acceptHeaderSupplier, TrustedDirectoryService trustedDirectoryService, 
+            RegistrationRequestBuilder registrationRequestBuilder, JwtDecoder jwtDecoder,
+            ResponseFactory responseFactory) {
+        Reject.ifNull(registrationEntitySupplier, "registrationEntitySupplier must be provided");
+        Reject.ifNull(acceptHeaderSupplier, "acceptHeaderSupplier must be provided");
+        Reject.ifNull(trustedDirectoryService, "trustedDirectoryService must be provided");
+        Reject.ifNull(registrationRequestBuilder, "registrationRequestBuilder must be provided");
+        Reject.ifNull(jwtDecoder, "jwtDecoder must be provided");
+        Reject.ifNull(responseFactory, "responseFactory must be provided");
+        this.registrationEntitySupplier = registrationEntitySupplier;
+        this.acceptHeaderSupplier = acceptHeaderSupplier;
+        this.trustedDirectoryService = trustedDirectoryService;
+        this.registrationRequestBuilder = registrationRequestBuilder;
+        this.jwtDecoder = jwtDecoder;
+        this.responseFactory = responseFactory;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        final String transactionId = FAPIUtils.getFapiInteractionIdForDisplay(context);
+        log.debug("({}) Running RegistrationRequestEntityValidatorFilter", transactionId);
+        try {
+            String b64EncodedRegistrationRequestEntity = this.registrationEntitySupplier.apply(context, request);
+            RegistrationRequest registrationRequest = this.registrationRequestBuilder.build(transactionId,
+                    b64EncodedRegistrationRequestEntity);
+            context.asContext(AttributesContext.class).getAttributes().put(REGISTRATION_REQUEST_KEY,
+                    registrationRequest);
+            log.info("({}) created context attribute " + REGISTRATION_REQUEST_KEY, transactionId);
+            return next.handle(context, request);
+        } catch (DCRException exception){
+            List<String> acceptHeader = acceptHeaderSupplier.apply(context, request);
+            Response response = responseFactory.getResponse(transactionId, acceptHeader, Status.BAD_REQUEST,
+                    exception.getErrorFields());
+            log.info("({}) Failed to understand the Registration Request body: {}", transactionId,
+                    exception.getMessage(), exception);
+            return Promises.newResultPromise(response);
```

Comment:
Sorry but I don't really see the use case for the content type negotiation for error responses.

Here are my concerns:

- [OAuth spec](https://www.rfc-editor.org/rfc/rfc7591#section-3.2.2) states that error messages are application/json
- Content negotiation is only done for the error path, if I were using an API I would expect it to apply to all responses
- I don't expect to see much usage of this feature, of the choices on offer json is the most popular.
  - Json is also easy to read by humans and machines (especially these simple error objects)
- It is adding extra complexity to an already large and complex PR

Appreciate that a lot of effort has gone in, especially to write tests but I really do question why we need this functionality.

I like that there is a common errorResponseFactory, this is an approach that we should continue to use to ensure consistent error responses are produced (and cut duplication).